Revised hotkey system

mairdl opened the pull request:
## Revised Hotkey System##
In the next release of JabRef we will introduce a revisid hotkey system. The 2 goals of this revision is to improve maneuverability through the keyboard and streamlining hotkeys. Some related hotkeys like for the side panes were scattered across the keyboard, so we grouped them together. You might be wondering what has changed, so have a look at them:

- Change default key bindings (have a look at the table below)
- Add a hotkey for creating a new technical report
- Add a hotkey for opening the OpenOffice/LibreOffice pane
- Add “Print preview” to the right-click menu
- Improve group pane maneuverability by making groups selectable by arrow keys 
- Improve main table maneuverability by making “TAB” jump to the next entry instead of the next cell, “SHIFT+TAB” selects the previous entry
- Resize the “change key bindings” dialog to 500x500


Keys - new | Keys - old | Function
-----------|------------|---------
F7	| alt + f | Automatically set file links
F8	| ctrl + shift + F7  | Cleanup entries
alt + 1	| ctrl + shift + E | Focus entry table
alt + 2	| ctrl + F9 | Toggle entry/preview editor 
alt + 3	| ctrl + shift + G | Toggle groups
alt + 4	| F5  | web search
alt + 0	|- | Open Openoffice/LibreOffice connection 
ctrl + shift + F7 | ctrl + f4 |	Synchronize file links 
ctrl + shift R | - | Technical report
- |alt + P | Print entry preview
- |ctrl + alt + T  | Hide/show toolbar
-|ctrl+P| Edit preamble

As you can see we moved related functions together. “F7” is now affiliated with file links, hence we changed key bindings if they contained “F7” but weren’t related to file links.
Also we moved all side panes together by changing them to “ALT+(1,2,3,4,0)”. By doing so we hope to improve the accessibility, because now you don’t have to remember all kinds of key combinations for side panes, only which number stands for your desired pane. On a side note: Open OpenOffice/LibreOffice connection was added to this family due to being a side pane.
The hotkey for “Edit preamble” was changed due to normally being the hotkey to print.
The new hotkey for “New technical report” was requested by some users that use it frequently, so we added a key binding.
We removed the hotkey for “Hide/show toolbar” due to being not of any imaginable use for users.
We removed the hotkey for “Print entry preview” due to some conflicts with others and the fact that this function wasn’t really used that often by users. But we didn’t remove it completely, we added it to the right click menu.
We removed the hotkey for "Edit preamble" due to being a seldomly used action that doens't require a hotkey.
We resized the “change key bindings” dialog due to being too small for some descriptions.

For questions about the functioning or suggestions for improvements please use the discussion forum. In case you have trouble, report bugs on GitHub.

mairdl commented:
# Key Binding Changes
Keys - new | Keys - old | Function
-----------|------------|---------
F7	| alt + f | Automatically set file links
F8	| ctrl + shift + F7  | Cleanup entries
alt + F7 | alt+ F | Automatically set file links
alt + 1	| ctrl + shift + E | Focus entry table
alt + 2	| ctrl + F9 | Toggle entry/preview editor 
alt + 3	| ctrl + shift + G | Toggle groups
alt + 4	| | Open office/libre office connection 
alt + 5	| F5 | web search
alt + 6	| ctrl + alt + T | hide/show toolbar  
ctrl + shift + F7 | ctrl + f4 |	Synchronize file links 
Ctrl + shift R |  | Technical report

## Deleted

P	- Print entry preview

boceckts commented:
I think you should leave a comment in the changelog

boceckts commented:
Please fix failing check.

oscargus commented:
I think that is currently the only way to print the preview. Unless one enables the preview toolbar, which, if I recall correctly, is only supported by manual editing of the preferences... But I may be wrong here.

mairdl commented:
I have disscussed the new bindings with @koppor. As far as I remember he said that print entry preview isn't needed at all.

tobiasdiez commented:
- Printing preview: just add it to the right-click menu
- I would put the websearch on Alt + 4 (more important then OpenOffice)
- for the same reason, OO -> Alt + 0
- Remove default Toggle Toolbar shortcut (can't imagine that this is useful for somebody, I would leave alt + num for toggling real panes)
- What is the difference between F7 and Alt + F7, both are mapped to the same command according to your table.

mairdl commented:
I fixed my table regarding Automatically set file links (F7 was the right one)

Braunch commented:
@mairdl are you done here? Because then we could remove the "[WIP]".

koppor commented:
Please also test how this works in conjunction with the emacs mode.

mairdl commented:
Ok I'll look into it.

koppor commented:
- [ ] tab in the main table: When I click on an entry and select tab, the next entry should be selected

koppor commented:
Better support of groups:

1. Open a bib file with groups and 10 entries.
2. Click on entry 10
3. Click on a group
4. Press cursor down

Effect: Cursor moves to the first entry in the entry table

Desired effect:
 - [ ] Next group is selected

Side comment:
 - [ ] When the group window has the focus, tab should jump to the first entry in the entry table.

tobiasdiez commented:
Maybe one should consider that Tab always cycles to the next frame if there is nothing to select in the current subwindow. For example, tab in entry list goes to entry preview / edit. Tab in some field in entry edit goes to next field, but tab in last field goes to group window. Tab in group window goes to next sidepane component and from there back to the entry list.

Braunch commented:
@koppor I don't think tab for the next entry is a convenient interaction. As I encountered when working with the customize entry type dialog, lists and tables are all navigated with the arrow keys, so the next enry is reached by pressing arrow-down. I think we should keep that consistent as it is right now. Do you agree?

mairdl commented:
@koppor tab in the main table, this is already done by the `DOWN` Key

mairdl commented:
I have to agree with @Braunch

mairdl commented:
@koppor Regarding the group behavoir: This be in another PR.

mairdl commented:
[]Printing preview: just add it to the right-click menu
[x]I would put the websearch on Alt + 4 (more important then OpenOffice)
[x]for the same reason, OO -> Alt + 0
[x]Remove default Toggle Toolbar shortcut (can't imagine that this is useful for somebody, I would leave alt + num for toggling real panes)
[x]What is the difference between F7 and Alt + F7, both are mapped to the same command according to your table.

koppor commented:
@Braunch Currently, when pressing tab, the next cell in the entry table is selected. One cannot do anything at a cell, can I? For instance, when pressing space on the url/doi file column, I would expect that the same behavior as when left-clicking is done. This currently doesn't happen. So, either jump to the next line via tab or do something meaningful when a single field in the table is focused.

koppor commented on CHANGELOG.md:
```diff
@@ -19,6 +19,7 @@ We refer to [GitHub issues](https://github.com/JabRef/jabref/issues) by using `#
 - [#1249](https://github.com/JabRef/jabref/issues/1249) Date layout formatter added
 - Added ISBN integrity checker
 - Added filter to not show selected integrity checks
+- [#1225](https://github.com/JabRef/jabref/issues/1225): Hotkey-system should be consistent
```
"is now consistent"

koppor replied:
Where does one find information (help page / blog post / ...) about what was changed and how the hotkeys are configured now?

mairdl replied:
I wanted to add the Information when everything was finalized, do you think i should add it beforehand?

koppor commented on src/main/java/net/sf/jabref/gui/JabRefFrame.java:
```diff
@@ -322,7 +322,6 @@
     private final Action toggleToolbar = enableToggle(new AbstractAction(Localization.menuTitle("Hide/show toolbar")) {
 
         {
-            putValue(Action.ACCELERATOR_KEY, Globals.getKeyPrefs().getKey(KeyBinding.HIDE_SHOW_TOOLBAR));
```
Why was that removed?

mairdl replied:
@tobiasdiez suggested this change because i quote: 

> Remove default Toggle Toolbar shortcut  (can't imagine that this is useful for somebody, I would leave alt + num for toggling real panes)

koppor commented on src/main/java/net/sf/jabref/gui/keyboard/KeyBinding.java:
```diff
@@ -25,7 +25,7 @@
     DELETE_ENTRY("Delete entry", Localization.lang("Delete entry"), "DELETE"),
     EDIT_ENTRY("Edit entry", Localization.lang("Edit entry"), "ctrl E"),
-    EDIT_PREAMBLE("Edit preamble", Localization.lang("Edit preamble"), "ctrl P"),
+    EDIT_PREAMBLE("Edit preamble", Localization.lang("Edit preamble"), "alt S"),
```
This change is not shown at your overview https://github.com/JabRef/jabref/pull/1525#issue-162928053

Furthermore, `p` sounds more like **p**reamble than `s`.. Why was it changed?

Furthermore, `alt`+ letter is more accessing menus. What if the menu shortcut in some other language is `s`?

mairdl replied:
I have to agree with you, I'll revert that change. To be honost I don't remember why that was changed.

koppor commented on src/main/java/net/sf/jabref/gui/keyboard/KeyBinding.java:
```diff
@@ -80,14 +79,15 @@
     UNMARK_ENTRIES("Unmark entries", Localization.lang("Unmark entries"), "ctrl shift M"),
-    WEB_SEARCH("Web search", Localization.lang("Web search"), "F5"),
-    WRITE_XMP("Write XMP", Localization.lang("Write XMP"), "ctrl F7"),;
+    WEB_SEARCH("Web search", Localization.lang("Web search"), "alt + 4"),
```
Why not `F5` any more? Maybe you should really write a blog post to motivate your changes.

mairdl replied:
We changed that so that all the sidepanes can be opened with `alt + Numberkey`, I think a blogpost is a good idea but I would wait untill the Github disucssion is over and eveeryone is in agreement.

koppor commented on src/main/java/net/sf/jabref/gui/keyboard/KeyBinding.java:
```diff
@@ -80,14 +79,15 @@
-    WRITE_XMP("Write XMP", Localization.lang("Write XMP"), "ctrl F7"),;
+    WEB_SEARCH("Web search", Localization.lang("Web search"), "alt + 4"),
+    WRITE_XMP("Write XMP", Localization.lang("Write XMP"), "F6"),
+    OPEN_OFFICE_LIBRE_CONNECTION("Open office/libre connection", Localization.lang("Open office/libre connection"), "alt 0"),;
```
The thin is called [LibreOffice](https://www.libreoffice.org/). Maybe use "Open/libre office connection". Have a look how it is handled in other places in JabRef.

oscargus replied:
"OpenOffice/LibreOffice" is the currently used "name", so "OpenOffice/LibreOffice connection" is available as a translation string already.

mairdl replied:
as @oscargus already mentioned `OpenOffice/LibreOffice` is the current used name, you can look that up in e.g. `JabRef_en.proberties`

mairdl commented:
@tobiasdiez 
> Printing preview: just add it to the right-click menu

Print preview is already a part of the right-click menu of the preview pane. Is that what you wanted me to do?

Braunch commented:
I am thinking about ways to handle multiple selected entries when clicking print entry preview within the right click menu. Should all the entries be printed or only the one the mouse was hovering over while clicking?

mairdl commented:
This is my work in progess blogpost. Feel free to leave some feedback. :+1: 

## Revised Hotkey System##
In the next release of JabRef we will introduce a reworked hotkey system. The 2 goals of this rework is to improve maneuverability through the keyboard and streamlining hotkeys. Some related hotkeys were scattered across the keyboard, so we grouped them together. You might be wondering what has changed, so have a look at them:

- Change default key bindings (have a look at the table below)
- Add a hotkey for creating a new technical report
- Add a hotkey for opening the OpenOffice/LibreOffice pane
- Add “Print preview” to the right-click menu
- Improve group pane maneuverability by making groups selectable by arrow keys 
- Improve main table maneuverability by making “TAB” jump to the next entry instead of the next cell, “SHIFT+TAB” selects the previous entry
- Resize the “change key bindings” dialog to 500x500


Keys - new | Keys - old | Function
-----------|------------|---------
F7	| alt + f | Automatically set file links
F8	| ctrl + shift + F7  | Cleanup entries
alt + 1	| ctrl + shift + E | Focus entry table
alt + 2	| ctrl + F9 | Toggle entry/preview editor 
alt + 3	| ctrl + shift + G | Toggle groups
alt + 4	| F5  | web search
alt + 0	|- | Open Openoffice/LibreOffice connection 
alt + p |ctrl + p| Edit preamble
ctrl + shift + F7 | ctrl + f4 |	Synchronize file links 
ctrl + shift R | - | Technical report
- |alt + P | Print entry preview
- |ctrl + alt + T  | Hide/show toolbar

As you can see we moved related functions together. “F7” is now affiliated with file links, hence we changed key bindings if they contained “F7” but weren’t related to file links.
Also we moved all side panes together by changing them to “ALT+(1,2,3,4,0)”. By doing so we hope to improve the accessibility, because now you don’t have to remember all kinds of key combinations for side panes, only which number stands for your desired pane. On a side note: Open OpenOffice/LibreOffice connection was added to this family due to being a side pane.
The hotkey for “Edit preamble” was changed due to normally being the hotkey to print.
The new hotkey for “New technical report” was requested by some users that use it frequently, so we added a key binding.
We removed the hotkey for “Hide/show toolbar” due to being not of any imaginable use for users.
We removed the hotkey for “Print entry preview” due to some conflicts with others and the fact that this function wasn’t really used that often by users. But we didn’t remove it completely, we added it to the right click menu.
We resized the “change key bindings” dialog due to being too small for some descriptions.

For questions about the functioning or suggestions for improvements please use the discussion forum. In case you have trouble, report bugs on GitHub.

mairdl commented:
So are there any more changes you want me to include?

mairdl commented:
I discovered a bug with the `PrintEntryPreview` action in the right click menu, it wouldn't work if the `EntryPreview` isn't active. So I changed my code to open the `EntryPreview` if it's not active.

Siedlerchr commented on src/main/java/net/sf/jabref/gui/BasePanel.java:
```diff
@@ -2142,11 +2142,15 @@ public void action() {
-         @Override
-         public void action() throws Throwable {
+        @Override
+        public void action() throws Throwable {
```
Is is possible to throw a more specific exception? Probably there are some other actions which have to be checked, too. 
It not then I would recommend to at least change it to Exception.

mairdl replied:
Sorry I don't understand what you want to say. What do you mean by "call cases"?

mairdl replied:
I checked some other implementations for `BaseAction`. The majority doesn't throw an excpetion. Some do throw `Throwable` or a specified exception.

Siedlerchr replied:
Throwing a throwable is usually a bad idea:
http://stackoverflow.com/questions/12564428/is-throws-throwable-good-practice
Change throwable to Exception, if no more specific exception is avaiable.

mairdl replied:
Done.

mairdl commented:
Codacy issues seem to be unrelated to this PR.

stefan-kolb commented:
I think edit preamble is kind of unnecessary.

mairdl commented:
Do you mean the hotkey or the function at whole?

stefan-kolb commented:
@mairdl The hotkey. The function is also not very useful probably, but the hotkey for it is even more unneeded for such a seldomly used action.

mairdl commented:
I'll remove it in that case.

mairdl commented:
Is there something else to add/delete/change?

koppor commented:
Does the "cycle" functionality for groups also work for the web search? (Focus if opened and not focused, close if focused and hotkey is pressed).

Braunch commented:
The toggle functionality is in a different PR: https://github.com/JabRef/jabref/pull/1605
and yes it does work there.

Braunch commented:
Are there any more changes to be done or can this be merged?

koppor commented:
Please resolve conflicts and prepare a blog post for https://github.com/JabRef/www.jabref.org (as pull request)

matthiasgeiger commented:
Side note: There is still a bug open regarding the key bindings (https://github.com/JabRef/jabref/issues/1235) - Perhaps you can have a look this as well?

mairdl commented:
I created a PR for my blogpost over at JabRef/www.jabref.org

Braunch commented:
Sieht nicht qus als hätte es mit uns zu tun

koppor commented:
Blog article at http://blog.jabref.org/2016/08/15/HotkeyRevision/
Currently, no help page on the hot keys.
